enhancements/update/targeted-update-edge-blocking.md

```diff
@@ -0,0 +1,966 @@
+	// +required
+	release Release `json:"release"`
+
+	// qualifiers represents the range of issues associated with
+	// updating to the target release. The cluster-version
+	// operator will evaluate all entries, and only recommend the
+	// update if there is at least one entry and all entries
+	// recommend the update.
+	// +optional
+	Qualifiers []ConditionalUpdateQualifier `json:"qualifiers,omitempty"`
+
+	// conditions represents the observations of the conditional update's
+	// current status. Known types are:
+	// * Recommended, for whether the update is recommended for the current cluster.
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type" protobuf:"bytes,1,rep,name=conditions"`
+}
+```
+
+The new ConditionalUpdateQualifier type will have the following schema:
+
+```go
+// ConditionalUpdateQualifier represents a reason and cluster-state
+// for not recommending a conditional update.
+// +k8s:deepcopy-gen=true
+type ConditionalUpdateQualifier struct {
+	// url contains information about this qualifier.
+	// +optional
+	URL URL `json:"url,omitempty"`
+
+	// reason is the CamelCase reason for not recommending a
+	// conditional update, in the event that qualifiers match the
+	// cluster state.
+	// +optional
+	Reason string `json:"reason,omitempty"`
+
+	// message provides additional information about the risk of
+	// updating, in the event that qualifiers match the cluster
+	// state. This is only to be consumed by humans. It may
+	// contain Line Feed characters (U+000A), which should be
+	// rendered as new lines.
+	// +optional
+	Message string `json:"message,omitempty"`
+
+	// matchingRules is a slice of conditions for deciding which
+	// clusters have the update recommended (happy conditions) and
+	// which do not (sad conditions). The slice is ordered by
+	// decreasing precedence. The cluster-version operator will
+	// walk the slice in order, and stop after the first it can
+	// successfully evaluate. If no condition can be successfully
+	// evaluated, the update will not be recommended.
+	// +optional
+	MatchingRules []ClusterCondition `json:"matchingRules,omitempty"`
+}
+```
+
+The new ClusterCondition type will have the following schema:
+
+```go
+// ClusterCondition represents an update which is recommended to some
+// clusters on the version the current cluster is reconciling, but which
+// may not be recommended for the current cluster.
+// +k8s:deepcopy-gen=true
+type ClusterCondition struct {
+	// type represents the cluster-condition type. This defines
+	// the members and semantics of any 'extra' properties.
+	// +required
+	Type string
+
+	// Extra represents unstructured, type-specific properties.
+	// +optional
+	Extra map[string]interface{}
+}
+```
+
+To work around [Go's lack of inline unions][go-union], we explicitly define the serialization:
```

Comment:
i don't understand this, we have implemented discriminated unions in our api w/o specalized serialization elsewhere, e.g.:

https://github.com/openshift/api/blob/85977bee07221f012896dcc53b77f44da0be0c4e/build/v1/types.go#L413-L437

Reply:
I've pushed 97f1ff4 with a new *Discriminating union implementation* section trying to outline some options and explain my choice.

Reply:
ok, a bunch more hashing out (mostly reflected in new *Alternatives* sections), and now we have a central, struct-pointer discriminated union in 6be15d990bb12c4b215ed17cf02e9b367bf495be.